StreamChatClient_v3/ChatClient.swift

```diff
@@ -0,0 +1,85 @@
+//
+// ChatClient.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+public final class ChatClient {
+  // MARK: - Public
+
+  public let currentUser: User
+
+  public let config: Config
+
+  public convenience init(currentUser: User, config: ChatClient.Config = .init()) {
+    // All production workers
+    let workers: [WorkerBuilder] = [
+      MessageSender.init
+    ]
+
+    self.init(
+      currentUser: currentUser,
+      config: config,
+      workers: workers,
+      environment: .init()
+    )
+  }
+
+  // MARK: - Internal
+
+  struct Environment {
+    var apiClientBuilder: (_ currentUser: User) -> APIClient = APIClient.init
+    var webSocketClientBuilder: (_ currentUser: User) -> WebSocketClient = WebSocketClient.init
+  }
+
+  var workers: [Worker]! // TODO: make `let` or `private(set) lazy var`
+
+  private let environment: Environment
+
+  private lazy var apiClient: APIClient = self.environment.apiClientBuilder(self.currentUser)
+
+  private lazy var webSocketClient: WebSocketClient = self.environment.webSocketClientBuilder(self.currentUser)
+
+  private lazy var persistentContainer: NSPersistentContainer = {
+    let container = NSPersistentContainer(name: "StreamChatModel")
+
+    let description = NSPersistentStoreDescription()
+    // TODO: Something like this?
+    if currentUser.isAnonymous || config.isLocalStorageEnabled == false {
+      // Use only in-memory store
+      description.url = URL(fileURLWithPath: "/dev/null")
```

Comment:
Why don't we use `NSInMemoryStoreType`?
https://developer.apple.com/documentation/coredata/nspersistentstore/persistent_store_types

Reply:
I wanted but then I found this bug https://stackoverflow.com/questions/56077716/core-data-of-type-nsinmemorystoretype-ignores-entitys-constraints . This solution is better, anyway. We use the same backing store for both types, the only difference is, one is persistent and the other not.